Get sub layer

ArneNx opened the pull request:
Added functionality to allow addressing of sub-layers vie "/" separated path. 
This is useful e.g. if you want to share parameters between two recurrent units.

albertz commented:
Hey, can you split this into multiple pull requests? Only keep the commits here which are really about the sub layer.

albertz commented on TFEngine.py:
```diff
@@ -954,10 +954,12 @@ def init_network_from_config(self, config=None, net_dict_post_proc=None):
         loader = CustomCheckpointLoader(
-          filename=model_filename, saveable_params=self.network.get_trainable_params(),
+          filename=model_filename,
+          saveable_params=self.network.get_params_list() if ignore_missing else self.network.get_trainable_params(),
```
Why? (But anyway, separate pull request.)

albertz commented on TFNetwork.py:
```diff
@@ -684,6 +692,7 @@ def get_layer(self, layer_name):
     :param str layer_name:
     :rtype: LayerBase
     """
+
```
Why?

albertz replied:
I mean the added empty line.

ArneNx replied:
Ah, ok. Fixed that.

albertz commented on TFNetwork.py:
```diff
@@ -696,6 +705,12 @@ def get_layer(self, layer_name):
       if not self.parent_net:
         raise LayerNotFound("cannot get layer %r, no parent net for %r" % (layer_name, self))
       return self.parent_net.get_layer(layer_name[len("base:"):])
+    if '/' in layer_name:
```
This logic should not be in `construct_layer`, but in `get_layer`.

ArneNx replied:
I guess you're talking about line 391 to 396, right? 
Until now, the `get_layer` method is not performing/ calling any construction. Do you really want to change that?

albertz replied:
Maybe this was moved. I think it was in `LayerBase.construct_layer` before. Now it's in `LayerBase.get_layer`, which is what I wanted.

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -720,6 +722,15 @@ def is_prev_step_layer(cls, layer):
 
+  def get_sub_layer(self, layer_name: str):
+    """
+    :param layer_name:  The sub_layer addressed by '/' separated path.
```
Define the type in the docstring: `:param str layer_name`

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -720,6 +722,15 @@ def is_prev_step_layer(cls, layer):
       return layer.is_prev_time_frame
     return False
 
+  def get_sub_layer(self, layer_name: str):
```
This is Python 3 only. We still want to keep Python 2 support (not sure for how long, but for now).

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -720,6 +722,15 @@ def is_prev_step_layer(cls, layer):
+  def get_sub_layer(self, layer_name: str):
+    """
+    :param layer_name:  The sub_layer addressed by '/' separated path.
+    :return: The sub_layer addressed in layer_name or None if no sub_layer exists
```
Also define the return type, i.e. like `:rtype: LayerBase|None`.

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -1059,6 +1070,19 @@ def get_layer(name):
+    """
+    :param layer_name:  The sub_layer addressed by '/' separated path.
+    :return: The sub_layer addressed in layer_name or None if no sub_layer exists
+    """
```
Same comments as before.

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -720,6 +722,15 @@ def is_prev_step_layer(cls, layer):
+    :param layer_name:  The sub_layer addressed by '/' separated path.
+    :return: The sub_layer addressed in layer_name or None if no sub_layer exists
+    """
+    if '/' in layer_name and layer_name.split('/')[0] == 'unit':
```
Why this logic? Why only if it is `unit`? I would not do that, but rather for any `layer_name`.
But instead, check `isinstance(self.cell, _SubnetworkRecCell)`, because only in that case you want that.

albertz commented on tests/test_TFNetworkRecLayer.py:
```diff
@@ -1429,7 +1429,6 @@ def train(net):
   assert not train_not_optim_out_layer.cell.output_layers_moved_out
   train(train_not_optim_net)
 
-
```
Why?

albertz commented:
Also many tests fail (I didn't really looked at the details). Please check.

albertz commented on TFNetwork.py:
```diff
@@ -1,6 +1,8 @@
 
 from __future__ import print_function
 
+import pprint
```
Do `from pprint import pprint` instead. Or better: Don't do that here at all. Do it locally in the method/function where you need it.

albertz commented on TFNetwork.py:
```diff
@@ -1,6 +1,8 @@
 from __future__ import print_function
 
+import pprint
+
```
No need for this extra empty line.

albertz commented on LmDataset.py:
```diff
@@ -84,6 +89,12 @@ def __init__(self,
       self.seq_gen = None
+    if orth_symbols_map_file and orth_symbols_map_file.endswith('.pkl'):
+      with open(orth_symbols_map_file, 'rb') as f:
+        self.orth_symbols_map = pickle.load(f)
```
Maybe the `import pickle` locally here.

albertz commented on LmDataset.py:
```diff
@@ -13,6 +13,7 @@
 import numpy
 import time
 import re
+import pickle
```
Maybe move below.

ArneNx commented:
I think I have addressed all your remarks now. 
The only thing remaining is our discussion about whether to put some of the logic into `construct_layer` (see my response above). Have you taken another look at this? 
And yes, this feature is definitely needed!

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -1067,6 +1081,20 @@ def get_layer(name):
       get_layer(layer_name)
       assert layer_name in self.net.layers
 
+  def get_sub_layer(self, layer_name):
```
Better rename to `get_layer`.

albertz replied:
Or maybe `get_layer_from_outside`.

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -1067,6 +1081,20 @@ def get_layer(name):
+    elif self.input_layers_net and layer_name in self.input_layers_net.layers:
+      return self.input_layers_net.layers[layer_name]
+    elif self.net and layer_name in self.net.layers:
+      return self.net.layers[layer_name]
```
I'm not sure if it is a good idea to allow access to `self.net` here as well. Because this is inside the recurrent loop, thus:
* It is without the time dim. This is not consistent when you get the layer from e.g. output_layers_net or input_layers_net. But it should be consistent.
* You also cannot access the `layer.output.placeholder`, because it is inside the recurrent loop. So depending on the use case, this doesn't make sense.

I think I would throw an exception here, like:
```suggestion
      raise Exception("%r: Cannot get layer %r from outside, because it is only available inside the recurrent loop. Add 'is_output_layer':True to the layer options." % (self.parent_rec_layer, layer_name))
```

albertz commented on TFNetwork.py:
```diff
@@ -388,6 +388,12 @@ def construct_layer(self, net_dict, name, get_layer=None, add_layer=None, check_
+        # but maybe it's not constructed yet, so try constructing the root layer
+        self.construct_layer(net_dict, name.split('/')[0], get_layer, add_layer, check_existing)
+        # constructing the root layer should have constructed all its children
+        return self.get_layer(name)  # ...so try again now
```
It should not call `self.construct_layer` in all cases, but rather `get_layer` (the param passed to this function).
Thus, move that code `if not get_layer: ...`, which is below, above this, and then call `get_layer` here.

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -725,6 +727,18 @@ def is_prev_step_layer(cls, layer):
+    :return: The sub_layer addressed in layer_name or None if no sub_layer exists
+    :rtype: LayerBase|None
+    """
+    if '/' in layer_name and isinstance(self.cell, _SubnetworkRecCell):
```
Why the extra `"unit/"`? I think this is confusing, and also more complicated than it needs to be.

albertz replied:
I would not do that. This is purely artificial. Actually I don't like it too much that it is called `"unit"` currently. Maybe we rename that later. But anyway, this logic is kind of independent (or should be independent) how the parameter of the class is named. So I would just remove that.

ArneNx replied:
I can see why this would be simpler (although I still think people will first try to access it through the path they see in the network dictionary).
So I would change this to remain independent of changes in naming of other layers. 
But what would you think of having something like 
```
if 'unit' == layer_name.split('.')[0]: 
  layer_name = layer_name.split('.')[1:] 
```

albertz replied:
Hm, I don't think so. I would just remove the `"unit"` part. It does not really add any new functionality or ability. And also keep it simple for now, so no extra cases.

ArneNx replied:
Ok... I changed it to expect "base_layer/sub_layer"

albertz commented on TFNetwork.py:
```diff
@@ -383,18 +383,24 @@ def construct_layer(self, net_dict, name, get_layer=None, add_layer=None, check_
+        # but maybe it's not constructed yet, so try constructing the root layer
+        get_layer(name.split('/')[0])
+        # constructing the root layer should have constructed all its children
+        return get_layer(name)  # ...so try again now
```
Do not call `get_layer(name)` here. This could lead to an infinite recursion (if this is not an existing hierarchical path).
Rather, assert that `name.split('/')[0] in self.layers` now, and then move on right here, so it will go into the code `if '/' in layer_name: ...` below.

ArneNx replied:
You're right. I ignored that case. 
See my newest change.

albertz commented on TFNetwork.py:
```diff
@@ -383,18 +383,26 @@ def construct_layer(self, net_dict, name, get_layer=None, add_layer=None, check_
+        # it may be a hierarchical path to a sub-layer, which should have been found by get_layer()
+        # but maybe it's not constructed yet, so try constructing the root layer
+        get_layer(name.split('/')[0])
+        assert name.split('/')[0] in self.layers, "Root %r layer was not constructed! " \
```
Small detail: Instead of `\` for splitting this into two lines, I prefer to use brackets, like `assert ...., ("Root ... ! "` newline `"Unable to ..." % (...))`.
(Despite personal taste, there is also one small advantage of this: When an exception would occur here, the code line in the stack trace would be the last line of the statement, i.e. in your case the `"Unable to ..." % ..`. When showing that statement alone in a debug stack trace, you might miss that this is part of another statement. E.g. better_exchook looks if the stack frame code line contains any non-matching `)`, and only then it would show also the preceding code line.)
